File: src/main/java/com/starshootercity/abilities/DependantAbility.java
package com.starshootercity.abilities;

import net.kyori.adventure.key.Key;
import org.jetbrains.annotations.NotNull;

public interface DependantAbility extends Ability {
    @NotNull Key getDependencyKey();

    default @NotNull DependencyAbility getDependency() {
        return AbilityRegister.dependencyAbilityMap.getOrDefault(getDependencyKey(), new PlaceholderDependencyAbility());
    }

    default DependencyType getDependencyType() {
        return DependencyType.REGULAR;
    }

    enum DependencyType {
        REGULAR,
        INVERSE
    }
}

File: core/src/main/kotlin/ru/turbovadim/abilities/DependantAbility.kt
package ru.turbovadim.abilities

import net.kyori.adventure.key.Key

interface DependantAbility : Ability {
    val dependencyKey: Key

    val dependency: DependencyAbility
        get() = AbilityRegister.dependencyAbilityMap.getOrDefault(this.dependencyKey, PlaceholderDependencyAbility())

    val dependencyType: DependencyType
        get() = DependencyType.REGULAR

    enum class DependencyType {
        REGULAR,
        INVERSE
    }
}
